client/src/actions/sharedActions.js

```diff
@@ -0,0 +1,29 @@
+// @flow
+import produce from 'immer'
+import {get, set} from 'lodash'
+
+import type {GenericAction, Path} from '../types/reduxTypes'
+import type {State} from '../state'
+
+// merges new data into destination Path
+// assumes a certain kind of data format - will need refinment if the API's different
+export const receiveData = (
+  path: Path,
+  data: Array<Object> | Object,
+  dataProviderRef: string
+): GenericAction<State, Array<Object> | Object> => ({
+  type: `Received data from ${dataProviderRef}`,
+  payload: data,
+  reducer: (state, data) =>
+    produce(state, (draft): void => {
+      const dataObject = Array.isArray(data)
+        ? data.reduce((obj, current) => {
+          obj[current.id] = current
+          return obj
+        }, {})
+        : {
+          [data.id]: data,
+        }
+      set(draft, path, Object.assign(get(draft, path) || {}, dataObject))
```

Comment:
set(draft, path, {...{get(draft, path) || {}}, ...dataObject})

Reply:
shouldn't be? ...dataObject first  and ...{get(... second so you overwrite the old data with the new

Reply:
once again ,we're using this https://github.com/mweststrate/immer

Reply:
sorry, but what has this change with immer? I think ```Object.assign(get(draft, path) || {}, dataObject)``` is equivalent to ```{...{get(draft, path) || {}}, ...dataObject}```

Reply:
well, appart from `{...{get(draft, path) || {}}}` not being valid syntax (let's assume it's `{...get(draft,path)}` and handle the `|| {}` elsewhere) one creates a new object and second modifies the one existing in draft

Reply:
but now that we're talking about this - the whole set is useless here - it should be just the object.assign part